components/iam/pkg/oidc/service.go

```diff
@@ -0,0 +1,142 @@
+	"net/http"
+	"net/url"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/gitpod-io/gitpod/common-go/log"
+	"github.com/go-chi/chi/v5"
+	"golang.org/x/oauth2"
+)
+
+type OIDCService struct {
+	Handler chi.Router
+
+	configsById      map[string]*OIDCClientConfig
+	verifierByIssuer map[string]*oidc.IDTokenVerifier
+	providerByIssuer map[string]*oidc.Provider
+}
+
+type OIDCClientConfig struct {
+	ID           string
+	Issuer       string
+	OAuth2Config *oauth2.Config
+	OIDCConfig   *oidc.Config
+}
+
+type OIDCStartParams struct {
+	State       string
+	Nonce       string
+	AuthCodeURL string
+}
+
+type OIDCAuthResult struct {
+	IDToken *oidc.IDToken
+}
+
+func NewOIDCService() *OIDCService {
+	var s OIDCService
+	s.configsById = make(map[string]*OIDCClientConfig)
+	s.verifierByIssuer = make(map[string]*oidc.IDTokenVerifier)
+	s.providerByIssuer = make(map[string]*oidc.Provider)
+	return &s
+}
+
+func (service *OIDCService) AddClientConfig(config *OIDCClientConfig) error {
+	if service.providerByIssuer[config.Issuer] == nil {
+		provider, err := oidc.NewProvider(context.Background(), config.Issuer)
+		if err != nil {
+			return errors.New("OIDC discovery failed: " + err.Error())
+		}
+		service.providerByIssuer[config.Issuer] = provider
+		service.verifierByIssuer[config.Issuer] = provider.Verifier(config.OIDCConfig)
+	}
+
+	config.OAuth2Config.Endpoint = service.providerByIssuer[config.Issuer].Endpoint()
+	service.configsById[config.ID] = config
+	return nil
+}
+
+func (service *OIDCService) GetStartParams(config *OIDCClientConfig) (*OIDCStartParams, error) {
+	// TODO(at) state should be a JWT encoding a redirect location
+	// Using a random string to get the flow running.
+	state, err := randString(32)
+	if err != nil {
+		return nil, errors.New("failed to create state")
+	}
+
+	nonce, err := randString(32)
+	if err != nil {
+		return nil, errors.New("failed to create nonce")
+	}
+
+	// Nonce is the single option passed on to configure the consent page ATM.
+	authCodeURL := config.OAuth2Config.AuthCodeURL(state, oidc.Nonce(nonce))
+
+	return &OIDCStartParams{
+		AuthCodeURL: authCodeURL,
+		State:       state,
+		Nonce:       nonce,
+	}, nil
+}
+
+func randString(size int) (string, error) {
+	b := make([]byte, size)
+	if _, err := io.ReadFull(rand.Reader, b); err != nil {
+		return "", err
+	}
+	return base64.RawURLEncoding.EncodeToString(b), nil
+}
+
+func (service *OIDCService) GetClientConfigFromRequest(r *http.Request) (*OIDCClientConfig, error) {
+	issuerParam := r.URL.Query().Get("issuer")
+	if issuerParam == "" {
+		return nil, errors.New("issuer param not specified")
+	}
+	issuer, err := url.QueryUnescape(issuerParam)
+	if err != nil {
+		return nil, errors.New("bad issuer param")
+	}
+	log.WithField("issuer", issuer).Trace("at GetClientConfigFromRequest")
+
+	for _, value := range service.configsById {
+		if value.Issuer == issuer {
+			return value, nil
+		}
+	}
+	return nil, errors.New("failed to find OIDC config for request")
```

Comment:
This needs to return more specific errors, as these need to be used to directly respond. If the caller does not specify the right arguments, it's a user error

Reply:
That's true. I'll add an issue for that.

We need to identify the relevant error modes, to get a good signal (without noise) to be able to render something meaningful:
* no params => 400
* issuer (or maybe client id) + unmatched => 404
* state (including client id) + unmatched => 404
* client config matched + unauthorized => 401
(that list is not complete)